chapter3/src/3_4_1/main.rs

```diff
@@ -0,0 +1,29 @@
+use std::io::{stdin, Error, ErrorKind, Read};
+
+fn main() {
+    loop {
+        let mut buffer = vec![0; 5];
+
+        match stdin().read(&mut buffer) {
```

Comment:
ここのパターンマッチングは

```suggestion
        let n = stdin().read(&mut buffer)?;
        if n == 0 {
                println!("EOF");
                break;
        }
        println!(
                    "size={} input'{}'",
                    size,
                    std::str::from_utf8(&buffer).unwrap()
        );
```

としておくほうがよさそうに思いました。

`read` 関数は Go のように `EOF` だった際にはエラーを返しません。

返り値が `Ok(0)` だったときに実質 `EOF` に到達したとみなせます。

rustdoc には

1. Reader が EOF に到達した
2. バッファがそもそも0バイトだった

のどちらかのケースを示すと書かれています。Rust 的な対応としてはしたがって、結果が 0 だったかどうかで EOF かどうかを判定しておいて、残りのエラーは検出時点で伝播させておいて済ませておく、がベストかなと思っています。

Reply:
こっちのほうがスッキリはするので書き換えたいとおもいます。